File: 1000 and 1 tasks/new-programming-skills/spiral-matrix.cs
public class Solution {
    // рамки, направления

    enum Directions {
        Right,
        Down,
        Left,
        Top
    }

    public IList<int> SpiralOrder(int[][] matrix) {
        var result = new List<int>();

        var X = matrix.Length;
        var Y = matrix[0].Length;

        var x = 0;
        var y = 0;
        var direction = Directions.Right;

        var minXIncluding = 0;
        var maxXIncluding = X - 1;
        var minYIncluding = 0;
        var maxYIncluding = Y - 1;

        result.Add(matrix[x][y]);

        while (result.Count() < X*Y) {
            if (direction == Directions.Right) {
                while (y < maxYIncluding) {
                    y++;
                    result.Add(matrix[x][y]);
                    if (y == maxYIncluding) {
                        minXIncluding++;
                    }
                }
                direction = Directions.Down;
            }
            else if (direction == Directions.Down) {
                while (x < maxXIncluding) {
                    x++;
                    result.Add(matrix[x][y]);
                    if (x == maxXIncluding) {
                        maxYIncluding--;
                    }
                }
                direction = Directions.Left;
            }
            else if (direction == Directions.Left) {
                while (y > minYIncluding) {
                    y--;
                    result.Add(matrix[x][y]);
                    if (y == minYIncluding) {
                        maxXIncluding--;
                    }
                }
                 direction = Directions.Top;
            }
            else if (direction == Directions.Top) {
                while (x > minXIncluding) {
                    x--;
                    result.Add(matrix[x][y]);
                    if (x == minXIncluding) {
                        minYIncluding++;
                    }
                }
                direction = Directions.Right;
            }
        }

        return result;
    }
}

File: 1000 and 1 tasks/new-top-150/spiral-matrix.java
import java.util.*;

public class Solution {

    public enum Direction {
        RIGHT,
        DOWN,
        LEFT,
        UP
    }

    public List<Integer> spiralOrder(int[][] matrix) {
        var lineCount = matrix.length;
        var columnCount = matrix[0].length;
        var n = lineCount * columnCount;

        var result = new ArrayList<Integer>();

        var left = 0;
        var right = columnCount - 1;
        var top = 0;
        var bottom = lineCount - 1;

        var state = Direction.RIGHT;
        while (result.size() < n) {
            switch (state) {
                case RIGHT:
                    for (var j = left; j <= right; j++) {
                        result.add(matrix[top][j]);
                    }
                    top++;
                    state = Direction.DOWN;
                    break;
                case DOWN:
                    for (var i = top; i <= bottom; i++) {
                        result.add(matrix[i][right]);
                    }
                    right--;
                    state = Direction.LEFT;
                    break;
                case LEFT:
                    for (var j = right; j >= left; j--) {
                        result.add(matrix[bottom][j]);
                    }
                    bottom--;
                    state = Direction.UP;
                    break;
                case UP:
                    for (var i = bottom; i >= top; i--) {
                        result.add(matrix[i][left]);
                    }
                    left++;
                    state = Direction.RIGHT;
                    break;
                default:
                    throw new RuntimeException();
            }
        }


        return result;
    }
}